draft-ietf-tls-tls13.md

```diff
@@ -5636,6 +5759,72 @@ application protocols separately ensuring that confidential
+transferring money) to be duplicated, thus harming the site or the
+user.  Attackers can also store and replay 0-RTT messages in order to
+re-order them with respect to other messages, which may harm
+unsuspecting applications.
+
+If data can be replayed a large number of times, additional attacks
+become possible. Specifically, attackers can use multiple replays to
+exploit information leakage via side channels such as timing network
+caches or measuring the speed of cryptographic operations. In addition, they may
+be able to overload rate-limiting systems. For further description of
+these attacks, see {{Mac17}}.
+
+Ultimately, servers have the responsibility to protect themselves
+against attacks employing 0-RTT data replication. The mechanism
+described in {{anti-replay}} are intended to
+prevent replay at the TLS layer do not provide complete protection
+against receiving multiple copies of client data.
+TLS 1.3 falls back to the 1-RTT
+handshake when the server does not have any information about the
+client, e.g., because it is in a different cluster which does not
+share state or because the ticket has been deleted as described in
+{{single-use-tickets}}. If the application layer protocol retransmits
+data in this setting, then it is possible for an attacker to induce a
+replay attack by sending the ClientHello to both the original cluster
+(which processes the data immediately) and another cluster which will
+fall back to 1-RTT and process the data upon application layer
+replay. The scale of this attack is limited by the client's
+willingness to retry transactions and therefore only allows a limited amount
+of duplication, with each copy appearing as a new connection at
+the server.
+
+If implemented correctly, the mechanisms described in
+{{single-use-tickets}} and {{client-hello-recording}} prevent a
+replayed ClientHello and its associated 0-RTT data from being accepted
+multiple times by any cluster with consistent state; for servers
+which limit the use of 0-RTT to one cluster for a single ticket, then a given
+ClientHello and its associated 0-RTT data will only be accepted once.
+However, if state is not completely consistent,
+then an attacker might be able to have multiple copies of the data be
+accepted during the replication window.
+Because they do not know the exact details of server behavior, clients
+MUST NOT send messages in early data which are not safe to have
+replayed and which they would not be willing to retry across multiple
+1-RTT connections.
+
+Application protocols MUST NOT use 0-RTT data without a profile that
+defines its use. That profile needs to identify which messages or
+interactions are safe to use with 0-RTT and how to handle the
+situation when the server rejects 0-RTT and falls back to 1-RTT.
+
+In addition, to avoid accidental misuse, TLS implementations MUST NOT
+enable 0-RTT (either sending or receiving) unless specifically
+requested by the application and SHOULD NOT automatically resend 0-RTT
```

Comment:
So, I think this is less a matter of should vs should not, and more of the matter of "impossible while making a working TLS implementation".  When the client retries the connection, exporter secrets are guaranteed to change, and other TLS-negotiated parameters can change too, so retrying a request can result in incorrect application-layer payload unless the implementation is minimal enough to not provide any of those facilities.